### ISSUE TITLE
Transmit other-player actions

### ISSUE THREAD

**Issue by bbqbaron** [jhmullen/feast-react#16]
We need to be able to notify client 1 that client 2 has done something to shared state. One bare-bones approach is to transmit serialized actions via sockets and just update the Redux store on the other side in the same way. For purposes of this ticket, maybe we can just implement that connection: receive action, dispatch to store.

Some followup tickets I imagine we'd need:
* Providing some meaningful feedback about the other player's action, so the game state doesn't just _bam_ change on you. One trivial MVP option might be a text log to the side: `Player1 bought Blah` etc.
* Concurrency/action locking. For now, wait your turn!
* A view of the other player's game state, like their board etc
* More than two players in the entire universe

**Comment by bbqbaron** [comment]
Naturally the other player's actions can't be reduced in _exactly_ the same way as ours, since we don't want to put their stuff in our deck etc.

**Comment by jhmullen** [comment]
@bbqbaron Could you give me a bit more help on understanding how actions are currently routed between the two sessions?  I see that this middleware wraps our redux store, and we have this generic action-catcher on index.js that dispatches to our store.  How do the contents of `/middleware` come into play?

The other player is going to have their party visible, which I think I'll be taking on next.  How would you suggest I map this double-owned concept (i.e., myParty vs YourParty, and the fact that your/my are the reverse to each respective player)?  I hope that question makes sense.

**Comment by bbqbaron** [comment]
@jhmullen Sure thing. Redux middleware receives every action, before reduce, in the order we provide it to `applyMiddleware`. Each middleware can do essentially whatever it wants as long as it calls the provided callback with _something_, which then goes to the next thing in place of the original action--where "next thing" is either the next middleware, or finally the reducer.

In particular, the `socketLogger` middleware emits actions by anyone (say, "Client 1") to the server exactly as it receives them. The server broadcasts all received actions to all sockets except the sending socket, meaning that the other player (Client 2) receives the same action that we dispatched.

On Client 2, the socket-consuming code tags that action as `otherPlayer` and dispatches it to the store. The reducer is wrapped with `routeForOtherPlayer`, which basically runs the _same_ reducer it would have run for Client 2's player, except because the action was by Client 1's player, it discards all player-specific state and only updates shared keys.

But! We now need all players' states on the store so we can render them! The most correct solution probably involves treating other players' states differently than our own, since the server needs to conceal some information and just say "Hey they have 5 cards in hand; that's all you get". However, for simplicity (and the fact that this is for testing), it's probably simpler for now to just have all players have everything?

What about a set of changes like this? If you're working on the other-player UI, which seems like a design hassle, I'm happy to execute these:

* take player-specific state keys--`hand`, `discard`, etc--and create an object that, if we had a type system, we'd call `Player`. 

* put `players: Player[]` on the store in place of the singular `hand` etc we have there now. (Or we can have an object with player names or ids on it, of course)

* update reducer to expect an action to have something like `playerId` on it, and have it only reduce state for that player

* when the server receives a connection, it changes its action-broadcaster function to tag the outgoing action with a player id

* when the client receives an action, it no longer tags with `otherPlayer`, instead using our newly updated reducer

* this does create one new corollary need, which is that we might need more explicit player join/register behavior. a client just "showing up" does weird stuff now that we're being clear about who's who! I don't want this to cascade into the eventual game-lobby behavior we'll eventually need, so I might start with a hardcoded state like "Which player are you? One or two?"

**Comment by bbqbaron** [comment]
@jhmullen  Just to avoid stepping on you, I was going to start this--cool?

**Comment by jhmullen** [comment]
@bbqbaron I'm sorry, I've been distracted with other projects lately.  I'd love for you to start this!  I had just come back to give your explanation a good read, and only just saw your secondary comment.  

Have you pushed forward on this or not yet?  I will work on other things until then.

**Comment by jhmullen** [comment]
@bbqbaron I've given your comment a good read now, thank you very much for breaking that all down for me!  

For now I think it's best to have a drop-in "player1/player2" selection.  I'm actually quite interested to see the barebones code for this, as network based state management is already a new concept for me so I've had to give the code a few passes already :)

Thanks again for the writeup and assistance

**Comment by bbqbaron** [comment]
Ahoy ahoy; sorry for delay. I did start on this and pushed a branch a few days ago; I can carry it the rest of the way and open a PR soon. Will hopefully get it in before the inevitable Christmas Outage.

**Comment by jhmullen** [comment]
Just to update this - I changed the loading of "starter cards" from a hard coded generation to a CSV backed table on the existing Google Sheet.

I'm just wondering to myself, will we need to do any "unique-ifying" of the starting card IDs to differentiate multiple players having identical starter decks?  Until now I have relied on IDs being unique.  If we have shared locations like "Trash" then the IDs of the starter cards will need to have additional slugs on them, like a `-p2` perhaps?
